nemo/backends/pytorch/actions.py

```diff
@@ -1356,29 +1369,25 @@ def train(
 
                 curr_tensors_to_optimize = training_loop[self.step % len(training_loop)][1]
                 final_loss = 0
-                nan = False
                 for tensor in curr_tensors_to_optimize:
                     if (
                         torch.isnan(registered_tensors[tensor.unique_name]).any()
                         or torch.isinf(registered_tensors[tensor.unique_name]).any()
                     ):
-                        if stop_on_nan_loss:
-                            raise ValueError('Loss is NaN or inf - exiting')
-                        logging.warning('Loss is NaN or inf')
-                        curr_optimizer.zero_grad()
-                        nan = True
-                        break
+                        if (
+                            (stop_on_nan_loss)
+                            or (self._optim_level not in AmpOptimizations)
+                            or (self._optim_level == Optimization.mxprO0)
+                        ):
+                            # Set flag here and terminate at next eval step (to avoid desync and hanging).
+                            self.nan_or_inf = True
+                            logging.warning('Loss is NaN or inf, will terminate at next eval step')
```

Comment:
Can you also print the step here?
Something like ('Loss is NaN or inf at %d, will terminate at next eval step', self.step)

Reply:
Done